File: zz_TradeZoneScript/MovementValues.cpp
//Script to write movement script values.

#include <iostream>
#include <vector>
#include <fstream>
#include <iomanip>
#include <string>

using namespace std;

string capitalizeWord(string strToConvert)
{
    string temp;
    char temp_char;
    temp_char = toupper(strToConvert[0]);
    strToConvert[0] = temp_char;
    //remove underscore if it exists and capitalize second word.
    for(unsigned int i=0;i<strToConvert.length();i++)
    {
        if (strToConvert[i] == '_')
        {
            strToConvert[i] = ' ';
            temp_char = toupper(strToConvert[i + 1]);
            strToConvert[i + 1] = temp_char;
        }
    }
    return strToConvert;
}

void MovementConnectionScriptValues ()
{
    string trade_zone;
    string zones_from_file;
    vector<string> tradeZones;
    ofstream fw("script_values.txt", ofstream::out);
    if (fw.is_open())
    {
        ifstream input_file;
        input_file.open("inputf.in");
        while (zones_from_file != "DONE")
        {
            input_file >> zones_from_file;
            if (zones_from_file != "PRINT")
            {
                tradeZones.push_back(zones_from_file);
            }
            else if (zones_from_file == "PRINT")
            {
                for (int i = 0; i < tradeZones.size(); i++)
                {
                    fw << "---------------------------------------" << endl;
                    fw << "#Trade Zone: " << capitalizeWord(tradeZones.at(i)) << endl;
                    fw << "---------------------------------------" << endl;
                    for (int z = 0; z < tradeZones.size(); z++)
                    {
                        if ( tradeZones.at(z) != tradeZones.at(i))
                        {
                            fw << tradeZones.at(i) << "_to_" << tradeZones.at(z) << "_svalue = {" << endl;
                            fw << "    value = 100 #Base Value" << endl;
                            fw << "    subtract = " << tradeZones.at(i) << "_transportation_svalue" << endl;
                            fw << "    subtract = " << tradeZones.at(z) << "_transportation_svalue" << endl;
                            fw << "}" << endl;
                        }  
                    }                    
                }
            }
        }
        tradeZones.clear();
        fw.close();
    }          
}

void MovementScriptValues ()
{
    string trade_zone;
    string regions_from_file;
    vector<string> regions;
    ofstream fw("script_values_2.txt", ofstream::out);
    if (fw.is_open())
    {
        ifstream input_file;
        input_file.open("input_all.in");
        while (regions_from_file != "DONE")
        {
            input_file >> regions_from_file;
            if (regions_from_file == "START")
            {
                input_file >> trade_zone;
            }
            else if (regions_from_file != "PRINT")
            {
                regions.push_back(regions_from_file);
            }
            else if (regions_from_file == "PRINT")
            {
                fw << trade_zone << "_transportation_svalue = {" << endl;
                fw << "    value = 0" << endl;
                for (int i = 0; i < regions.size(); i++)
                {
                    fw << "    region:" << regions.at(i) << " = {" << endl;
                    fw << "        limit = {" << endl;
                    fw << "            any_region_province = {" << endl;
                    fw << "                OR = {" << endl;
                    fw << "                    any_neighbor_province = { has_road_towards = PREV }" << endl;
                    fw << "                    has_building = port_building" << endl;
                    fw << "                }" << endl;
                    fw << "            }" << endl;
                    fw << "        }" << endl;
                    fw << "        every_region_province = {" << endl;
                    fw << "            limit = {" << endl;
                    fw << "                any_neighbor_province = { has_road_towards = PREV }" << endl;
                    fw << "            }" << endl;
                    fw << "            add = railway_bonus" << endl;
                    fw << "        }" << endl;
                    fw << "        every_region_province = {" << endl;
                    fw << "            limit = {" << endl;
                    fw << "                has_building = port_building" << endl;
                    fw << "            }" << endl;
                    fw << "            add = port_bonus" << endl;
                    fw << "        }" << endl;
                    fw << "    }" << endl;
                }
                fw << "}" << endl;
            }
            if (regions_from_file == "CLEAR")
            {
                regions.clear();
            }
        }
        regions.clear();
        fw.close();
    }      
}

int main()
{
    MovementConnectionScriptValues();
    MovementScriptValues();
}

File: zz_TradeZoneScript/MovementValues.py
import TradeZones #Import the trade zones dictionary and list.

def print_trade_zone_header(trade_zone):
	caps_name = trade_zone.replace("_", " ").title()
	trade_zone_header = "\n#---------------------------------------\n#Trade Zone: {caps_name}\n#---------------------------------------\n".format(caps_name=caps_name)
	print(trade_zone_header)

def print_trade_zone_connection_values(trade_zone, trade_zone_2):
	zone_1 = trade_zone.lower()
	zone_2 = trade_zone_2.lower()
	trade_zone_connections = "{zone_1}_to_{zone_2}_svalue = {{\n    value = 100 #Base Value\n    subtract = {zone_1}_transportation_svalue\n    subtract = {zone_2}_transportation_svalue\n}}".format(zone_1=zone_1,zone_2=zone_2)
	print(trade_zone_connections)

def print_movement_script_values(trade_zone):
	print_trade_zone_header(trade_zone)
	print("{trade_zone}_transportation_svalue = {{\n    value = 0".format(trade_zone=trade_zone))
	for x in TradeZones.trade_zones_dict.get(trade_zone):
		for i in x:
			print("    region:{i} = {{\n        if = {{\n            limit = {{\n                any_region_province = {{\n                    OR = {{\n                        any_neighbor_province = {{ has_road_towards = PREV }}\n                        has_building = port_building\n                    }}\n                }}\n            }}\n            every_region_province = {{\n                limit = {{\n                    any_neighbor_province = {{ has_road_towards = PREV }}\n                }}\n                add = railway_bonus\n            }}\n            every_region_province = {{\n                limit = {{\n                    has_building = port_building\n                }}\n                add = port_bonus\n            }}\n        }}\n    }}".format(i=i))
	print("}")

def connection_values(trade_zone_list):
	for zone in trade_zone_list:
		print_trade_zone_header(zone)
		for regions in trade_zone_list:
			if ( regions != zone ):
				print_trade_zone_connection_values(zone, regions)
	#NOTE: This is the function that needs fixed. It currently prints connections that are repetitions but idk how to fix it yet.


#FUNCTION 1: Purpose - Print all base movement values
for trade_zone in TradeZones.trade_zones_dict:
  print_movement_script_values(trade_zone)

#FUNCTION 2: Purpose - Print all conncection movement values
#connection_values(TradeZones.trade_zones)